File: go/internal/handler/product_handler.go
package handler

import (
	"net/http"

	"github.com/app/gin-postgres-api/dto"
	"github.com/app/gin-postgres-api/internal/service"
	"github.com/app/gin-postgres-api/pkg/response"
	"github.com/gin-gonic/gin"
)

type ProductHandler interface {
	GetAll(c *gin.Context)
	GetByID(c *gin.Context)
	Create(c *gin.Context)
	Update(c *gin.Context)
	Delete(c *gin.Context)
}

type ProductHandlerImpl struct {
	productService service.ProductService
}

func NewProductHandler(productService service.ProductService) ProductHandler {
	return &ProductHandlerImpl{productService: productService}
}

func (h *ProductHandlerImpl) GetAll(c *gin.Context) {
	products, err := h.productService.GetAll()
	if err != nil {
		response.Error(c, http.StatusInternalServerError, err.Error())
		return
	}
	response.Success(c, http.StatusOK, products)
}

func (h *ProductHandlerImpl) GetByID(c *gin.Context) {
	product, err := h.productService.GetByID(c.Param("id"))
	if err != nil {
		response.Error(c, http.StatusNotFound, err.Error())
		return
	}
	response.Success(c, http.StatusOK, product)
}

func (h *ProductHandlerImpl) Create(c *gin.Context) {
	var body dto.ProductRequest

	if err := c.ShouldBindJSON(&body); err != nil {
		response.Error(c, http.StatusBadRequest, "name and price are required")
		return
	}

	userID, _ := c.Get("username")

	product, err := h.productService.Create(body.Name, body.Desc, body.Price, userID.(string))
	if err != nil {
		response.Error(c, http.StatusInternalServerError, err.Error())
		return
	}

	response.Success(c, http.StatusCreated, product)
}

func (h *ProductHandlerImpl) Update(c *gin.Context) {
	var body dto.ProductRequest

	if err := c.ShouldBindJSON(&body); err != nil {
		response.Error(c, http.StatusBadRequest, err.Error())
		return
	}

	product, err := h.productService.Update(c.Param("id"), body.Name, body.Desc, body.Price)
	if err != nil {
		status := http.StatusInternalServerError
		if err.Error() == "product not found" {
			status = http.StatusNotFound
		}
		response.Error(c, status, err.Error())
		return
	}

	response.Success(c, http.StatusOK, product)
}

func (h *ProductHandlerImpl) Delete(c *gin.Context) {
	if err := h.productService.Delete(c.Param("id")); err != nil {
		status := http.StatusInternalServerError
		if err.Error() == "product not found" {
			status = http.StatusNotFound
		}
		response.Error(c, status, err.Error())
		return
	}
	response.Success(c, http.StatusOK, nil)
}

File: php/app/Handlers/ProductHandler.php
<?php

declare(strict_types=1);

namespace App\Handlers;

use App\Services\ProductService;
use Psr\Http\Message\ResponseInterface as Response;
use Psr\Http\Message\ServerRequestInterface as Request;

class ProductHandler
{
    public function __construct(private ProductService $productService) {}

    public function getAll(Request $request, Response $response): Response
    {
        $products = $this->productService->getAll();
        return $this->json($response, ['success' => true, 'data' => $products]);
    }

    public function getById(Request $request, Response $response, array $args): Response
    {
        $product = $this->productService->getById($args['id']);

        if (!$product) {
            return $this->json($response, ['success' => false, 'message' => 'Product not found'], 404);
        }

        return $this->json($response, ['success' => true, 'data' => $product]);
    }

    public function create(Request $request, Response $response): Response
    {
        $body = (array) $request->getParsedBody();

        if (empty($body['name']) || !isset($body['price'])) {
            return $this->json($response, ['success' => false, 'message' => 'name and price are required'], 400);
        }

        // Get user_id from JWT payload stored in request attribute
        $token  = $request->getAttribute('token');
        $userId = $token->username ?? 'unknown';

        $product = $this->productService->create($body, $userId);

        return $this->json($response, ['success' => true, 'data' => $product], 201);
    }

    public function update(Request $request, Response $response, array $args): Response
    {
        $body = (array) $request->getParsedBody();

        $updated = $this->productService->update($args['id'], $body);

        if (!$updated) {
            return $this->json($response, ['success' => false, 'message' => 'Product not found'], 404);
        }

        $product = $this->productService->getById($args['id']);
        return $this->json($response, ['success' => true, 'data' => $product]);
    }

    public function delete(Request $request, Response $response, array $args): Response
    {
        $deleted = $this->productService->delete($args['id']);

        if (!$deleted) {
            return $this->json($response, ['success' => false, 'message' => 'Product not found'], 404);
        }

        return $this->json($response, ['success' => true, 'data' => null]);
    }

    private function json(Response $response, array $data, int $status = 200): Response
    {
        $response->getBody()->write(json_encode($data));
        return $response->withHeader('Content-Type', 'application/json')->withStatus($status);
    }
}
